agent/agent.go

```diff
@@ -828,6 +828,17 @@ func (a *Agent) runPostOrTeardownTaskCommands(ctx context.Context, tc *taskConte
 	ctx, span := a.tracer.Start(ctx, "post-task-commands")
 	defer span.End()
 
+	// We run the command cleanups in a defer in case any of the post commands add cleanups.
+	defer func() {
```

Comment:
How is `github.generate_token` intended to work when a task is in a task group? For task groups, is it valid to call `github.generate_token` from setup group or teardown group? If it was called in setup group, the first task that runs would end up deleting the generated token when it finishes, and prevent later tasks in the task group from using that shared setup. At the opposite end of the task group, it also means calling `github.generate_token` in teardown group would not have the token auto-revoked when it finishes.

Would be worth calling out in the task group docs when we document this command that it can't be used in a setup group/teardown group. Either that, or adding validation for setup/teardown group commands to disallow `github.generate_token`.

Reply:
Hm that's a good question, it's not something we considered in the design. We only planned to 'revoke the token when the task completes'. I would say supporting task groups persisting the same token between tasks is not in scope and add to the docs the 'workaround'.

I'm also not that sure how I would go about actually supporting persisting the expansion/token between multi host task groups. I would say that the hour lifespan for the tokens also will force users to sometimes generate multiple if they have a very long task, which is the same workaround

Reply:
After investigating some more, I believe the recent changes I just added support task group's setup and teardown. I manually tested on staging and the tokens worked before this change, and after the tokens all got revoked (ran generate token in each (setup|teardown)_(group|task) and tried using each token.

Reply:
Could we have a potentially simpler implementation where we run cleanup at the end of a task for a normal task, and the end of a task group (single-host or multi-host)? If so, all we would have to do is run the cleanup functions in a defer at the top of in `runTeardownGroupCommands` - that function runs at the end of every normal task (weird I know, but it is what it is) and for task groups, it runs at the end of the entire task group. We've used defers in `runTeardownGroupCommands` for that kind of purpose before.

Reply:
Just say this one, let me try this!